File: SegParte_Aula05/ClientServer/Backend/Simulacao.ts
import { format } from 'date-fns'; // Biblioteca Node.js para formatação de data

export class Simulacao {
    private readonly dataSimulacao: Date;
    private readonly valorEmprestimo: number;
    private readonly numParcelas: number;
    private readonly taxaJuros: number;
    private valorParcela: number = 0;
    private valorTotalPagar: number = 0;

    constructor(valorEmprestimo: number, numParcelas: number, taxaJuros: number) {
        this.dataSimulacao = new Date();
        this.valorEmprestimo = valorEmprestimo;
        this.numParcelas = numParcelas;
        this.taxaJuros = taxaJuros;
    }

    // Getters
    public getValorEmprestimo(): number { return this.valorEmprestimo; }
    public getNumParcelas(): number { return this.numParcelas; }
    public getTaxaJuros(): number { return this.taxaJuros; }
    public getValorParcela(): number { return this.valorParcela; }
    public getValorTotalPagar(): number { return this.valorTotalPagar; }

    // Setters
    public setValorParcela(valorParcela: number): void { this.valorParcela = valorParcela; }
    public setValorTotalPagar(valorTotalPagar: number): void { this.valorTotalPagar = valorTotalPagar; }

    // Método para representação em String (para salvar no arquivo)
    public toString(): string {
        const dataFormatada = format(this.dataSimulacao, 'dd/MM/yyyy HH:mm:ss');
        return `
            Data: ${dataFormatada}
            Empréstimo: R$ ${this.valorEmprestimo.toFixed(2)}, Parcelas: ${this.numParcelas}, Taxa: ${(this.taxaJuros * 100).toFixed(2)}%
            Valor Parcela: R$ ${this.valorParcela.toFixed(2)}, Total Pago: R$ ${this.valorTotalPagar.toFixed(2)}
        `;
    }
}

File: SegParte_Aula05/MVC/Simulacao.java
package MVC;

import java.time.LocalDateTime;

// Model: Entidade que representa os dados
public class Simulacao {
    private final LocalDateTime dataSimulacao;
    private final double valorEmprestimo;
    private final int numParcelas;
    private final double taxaJuros;
    private double valorParcela;
    private double valorTotalPagar;

    public Simulacao(double valorEmprestimo, int numParcelas, double taxaJuros) {
        this.dataSimulacao = LocalDateTime.now();
        this.valorEmprestimo = valorEmprestimo;
        this.numParcelas = numParcelas;
        this.taxaJuros = taxaJuros;
    }

    // Getters e Setters (Omissos por brevidade, mas essenciais em produção)
    public double getValorEmprestimo() { return valorEmprestimo; }
    public int getNumParcelas() { return numParcelas; }
    public double getTaxaJuros() { return taxaJuros; }
    public double getValorParcela() { return valorParcela; }
    public double getValorTotalPagar() { return valorTotalPagar; }
    public LocalDateTime getDataSimulacao() { return dataSimulacao; }

    public void setValorParcela(double valorParcela) { this.valorParcela = valorParcela; }
    public void setValorTotalPagar(double valorTotalPagar) { this.valorTotalPagar = valorTotalPagar; }

    @Override
    public String toString() {
        return String.format(
            "Data: %s\nEmpréstimo: R$ %.2f, Parcelas: %d, Taxa: %.2f%%\n" +
            "Valor Parcela: R$ %.2f, Total Pago: R$ %.2f\n",
            dataSimulacao.format(java.time.format.DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")),
            valorEmprestimo, numParcelas, taxaJuros * 100, valorParcela, valorTotalPagar
        );
    }
}